File: 104.二叉树的最大深度.js
/*
 * @lc app=leetcode.cn id=104 lang=javascript
 *
 * [104] 二叉树的最大深度
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxDepth = function(root) {
    return maxDepth_recursive(root);
};
// @lc code=end

// Recursion
var maxDepth = function(root) {
    if (root === null) {
        return 0;
    }
    return Math.max(maxDepth(root.left), maxDepth(root.right)) + 1;
};


// Loop
var maxDepth = function(root) {
    let level = 0;
    if (root === null) {
        return level;
    }
    let arr = [root];
    while (arr.length) {
        level++;
        let arr_temp = [];
        for (let node of arr) {
            if (node.left) {
                arr_temp.push(node.left);
            }
            if (node.right) {
                arr_temp.push(node.right);
            }
        }
        arr = arr_temp;
    }
    return level;
};
/*
var maxDepth = function(root) {
    let res = 0
    if (root == null) {
        return res;
    }
    let stack = [[root, 1]];
    while (stack.length > 0) {
        let [node, depth] = stack.pop();
        res = Math.max(res, depth);
        if (node.left) {
            stack.push([node.left, depth + 1])
        }
        if (node.right) {
            stack.push([node.right, depth + 1])
        }
    }
    return res;
};
*/

File: 104.二叉树的最大深度.rs
/*
 * @lc app=leetcode.cn id=104 lang=rust
 *
 * [104] 二叉树的最大深度
 */
struct Solution{}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
  pub val: i32,
  pub left: Option<Rc<RefCell<TreeNode>>>,
  pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
  #[inline]
  pub fn new(val: i32) -> Self {
    TreeNode {
      val,
      left: None,
      right: None
    }
  }
}

// @lc code=start
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::rc::Rc;
impl Solution {
    pub fn max_depth(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        if root.is_none() {
            return 0;
        }

        let mut depth = 0;
        let mut deque = std::collections::VecDeque::new();
        deque.push_back(root);

        while !deque.is_empty() {
            let level_size = deque.len();
            // 遍历新的一层，深度加1
            depth += 1;

            // 层级遍历，当前层节点弹出队列，同时将其左、右节点压入队列
            for _ in 0..level_size {
                if let Some(Some(node)) = deque.pop_front() {
                    if node.borrow().left.is_some() {
                        deque.push_back(node.borrow().left.clone());
                    }
                    if node.borrow().right.is_some() {
                        deque.push_back(node.borrow().right.clone());
                    }
                }
            }
        }
        depth
    }
}
// @lc code=end
